tests/test_utils/run_server.py

```diff
@@ -69,55 +69,69 @@ def make_dummy_server(host='0.0.0.0', port=None, num_experts=1, expert_cls='ffn'
 
 
 @contextmanager
-def background_server(*args, verbose=True, **kwargs):
-    """ Runs server in a background process and returns a reference to it. """
-    recv_addr, send_addr = mp.Pipe(duplex=True)
-    trigger_shutdown = mp.Event()
+def background_server(*args, shutdown_timeout=5, verbose=True, **kwargs):
+    """ A context manager that creates server in a background thread, awaits .ready on entry and shutdowns on exit """
+    pipe, runners_pipe = mp.Pipe(duplex=True)
+    runner = mp.get_context("spawn").Process(
+        target=_server_runner, args=(runners_pipe, *args), kwargs=dict(verbose=verbose, **kwargs))
 
-    def server_runner():
+    try:
+        runner.start()
+        yield pipe.recv()  # once the server is ready, runner will send us a tuple(hostname, port, dht port)
+        pipe.send('SHUTDOWN')  # on exit from context, send shutdown signal
+    finally:
         try:
-            server = make_dummy_server(*args, verbose=verbose, start=True, **kwargs)
-            dht_port = server.dht.port if server.dht is not None else None
-            send_addr.send((server.addr, server.port, dht_port))
-            trigger_shutdown.wait()
+            runner.join(timeout=shutdown_timeout)
         finally:
             if verbose:
-                print("Shutting down server...")
-            trigger_shutdown.set()  # if server failed internally, set the shutdown trigger anyway
-            server.shutdown()
+                print("Server failed to shutdown gracefully, terminating it the hard way...")
+            runner.terminate()
             if verbose:
-                print("Server shut down successfully.")
+                print("Server terminated.")
 
-    try:
-        runner = mp.Process(target=server_runner)
-        runner.start()
-        yield recv_addr.recv()  # yield tuple(hostname, port)
 
+def _server_runner(pipe, *args, verbose, **kwargs):
+    server = make_dummy_server(*args, verbose=verbose, start=True, **kwargs)
+    try:
+        dht_port = server.dht.port if server.dht is not None else None
+        pipe.send((server.addr, server.port, dht_port))
+        pipe.recv()  # wait for shutdown signal
     finally:
-        trigger_shutdown.set()
-        runner.terminate()
-        runner.join()
+        if verbose:
+            print("Shutting down server...")
+        server.shutdown()
+        if verbose:
+            print("Server shut down successfully.")
 
 
 if __name__ == '__main__':
```

Comment:
By the way, do we actually call run_server.py somewhere? I'm not sure this is the intended use for something located in `tests/test_utils`. If this is an example server launch script (which is used for debugging etc.), it should probably be located somewhere like `scripts/`

Reply:
i routinely call python -m test_utils.run_server.py for prototyping, it will also be useful in tutorials. I agree that it should be moved out of test utils, but this is a topic for another pr.

Reply:
Okay, but we'd better write it down somewhere